KIPs/kip-113.md

```diff
@@ -0,0 +1,252 @@
+- kip113ContractAddress: the address of a contract that implements IKIP_113.
+
+#### Validation
+
+After reading from Registry, the node must perform the following validations:
+
+- [KeyValidate](https://www.ietf.org/archive/id/draft-irtf-cfrg-bls-signature-05.html#section-2.5)
+- [PopVerify](https://www.ietf.org/archive/id/draft-irtf-cfrg-bls-signature-05.html#section-3.3.3)
+
+## Rationale
+
+### Design decisions
+
+#### Registry is not upgradeable
+
+The upgradeable proxy pattern allows the functionality of contracts to be upgraded without changing their addresses.
+However, Registry should not be implemented via upgradeable proxy contract due to its impact on the ecosystem:
+
+- Upgrade can be done covertly by the contract owner.
+- Upgrade can happen at an arbitrary block.
+- Upgrade key can be stolen.
+
+In other words, should Registry be upgraded, Registry must be redeployed to have a different address.
+
+#### Propagating the Registry address via ChainConfig
+
+There are several methods to propagate the a system contract address to the network:
+
+1. HeaderGov: deploy the contract from EOA, and make a governance vote from the governing node (the method used for Baobab GovParam address).
+2. ContractGov: deploy the contract from EOA, and make an on-chain governance vote.
+3. SpareContract: deploy the contract from EOA, and set the address to `AddressBook.spareContract`.
+4. ChainConfig: deploy the contract from EOA, and set the address to ChainConfig (the method used for TreasuryRebalance, a.k.a. [KIP-103](https://kips.klaytn.foundation/KIPs/kip-103)).
+5. Precompiled-contract: deploy the contract from EOA, and introduce a new Precompiled contract that returns its address.
+6. Bytecode-injection: inject a bytecode to a predefined address at hardfork.
+
+We can view each method from the following point of view:
+
+##### Explicit contract logic upgrade
+
+| Methods                                                | [1] HeaderGov | [2] ContractGov                        | [3] SpareContract | [4] ChainConfig | [5] Precompiled-contract | [6] Bytecode-injection |
+| ------------------------------------------------------ | ------------- | -------------------------------------- | ----------------- | --------------- | ------------------------ | ---------------------- |
+| Upgrade with hardfork                                  | X             | X                                      | X                 | O               | O                        | O                      |
+| Managing address change timing                         | O             | O                                      | X                 | O               | O                        | O                      |
+| Awareness of subscriber contracts after address update | X             | △<br> Latest GovParam address required | O                 | X               | O                        | X                      |
```

Comment:
Awareness of subscriber contracts after address update
-> This means that the subscriber contracts will notify after the address update?

Reply:
The question is: will the subscriber contracts be able to use the latest contract address?
e.g., for [3] spare contract, subscriber contracts can use `address registry = IAddressBook(0x400).spareContract()`

Reply:
I suggested the phrase in another comment.